### DISCUSSION TITLE
how to easily exclude properties from record equality?

### DISCUSSION THREAD

**Discussion by juepiezhongren** [dotnet/csharplang#4257] [Q&A]
i have a record
```
    public record Order(string UserId, DateTime Time) {
        public User User { get; set; }
    }
```
where User is not a init property and i dont want it to be within equality consideration.

any easy way to make it happen?

**Comment by theunrepentantgeek** [comment]
I believe, by design, the answer here is to implement `Equals()` yourself since you don't want the auto-implemented behaviour.

**Reply by 333fred** [reply]
This is correct. We believe that manual implementation and/or source generators will cover the scenario.

**Comment by juepiezhongren** [comment]
@theunrepentantgeek ，i'm also trying to do a source generator for my senarios, where simply a equal override seems not enough

**Comment by orthoxerox** [comment]
I agree that it's one of the logical growth points for this feature. Reimplementing the whole equality is usually a big pain in the neck if all you need is to add a sequence-equal collection member or exclude a volatile member.

An attribute that supplies a different equality comparer for a member would work well, but I would settle for a helper like `System.HashCode` (`System.RecordEquals`?)

**Reply by juepiezhongren** [reply]
recordEqual is really gr8

**Comment by chtenb** [comment]
Like others suggested, the intended way to solve this is by manually implementing the Equals method. If writing the Equals manually is really too cumbersome in your case or sensitive to mistakes, you can also consider wrapping the member in a type that implements equality as always true.

```csharp
public readonly struct IgnoreEquals<T>(T value) {
    public T Value { get; init; } = value;
    public override bool Equals(object? obj) => true;
    public override int GetHashCode() => 0;
    public override string? ToString() => Value?.ToString();
}
```

Then in your record types:

```csharp
record Foo {
    private IgnoreEquals<int> _ignoreThisMemberInEquals = new(0);
}
```

This way the compiler generated Equals will still work.
Of course the price you have to pay is that you have to type `.Value` everytime you want to access the value.

**Comment by Audionysos** [comment]
This gives me WPF vibes were when you want to change a simple button color you need to rewrite 200 lines of style/template. I'm so disappointed :(

**Comment by Audionysos** [comment]
Ok, this is stupid, may not be the fastest, but if anyone is interested in different workaround, this works and it's least boilerplate solution I could think of.
The idea is basically you have a single static object for your record type which track actual references and use this object in property accessors. Something like this:

```csharp
public class DynamicStorage<T> {
	private Dictionary<long, Props> dic = [];
	private ObjectIDGenerator g = new ObjectIDGenerator();
	public Props this[T tx] {
		get {
			var t = g.GetId(tx, out var ft);
			if(dic.TryGetValue(t, out var p)) return p;
			dic.Add(t, p = new Props());  return p;
		}
	}
}

public class Props {
	private Dictionary<string, object?> dic = [];
	public T? get<T>(string name) => dic.TryGetValue(name, out var v)
		? (T?)v : default;
	public void Set<T>(string name, T? v) => dic[name] = v;
}
```

Then you use it like this:

```csharp
public record class TargetFramework  (
	TargetFrameworkName? name,
	Version? version,
	string? platform = null)
{
	private static DynamicStorage<TargetFramework> store = new();
	
	/// <summary>Raw string stored in `TargetFramework(s)` or `TargetFrameworkVersion`.
	/// PS. The stupid records don't allow to exclude properties form equality check so currently original "v4.8" would be converted to "net48" for the operators to work as expected.</summary>
	public string? sourceString {
		get => store[this].get<string>(nameof(sourceString));
		private set => store[this].Set(nameof(sourceString), value);
	}

//...
```

This will allows me to have two different `sourceString` values for two equal records:

```csharp
var t1 = (TargetFramework)"net48";
var t2 = (TargetFramework)"v4.8";
Console.WriteLine(t1.sourceString); //net48
Console.WriteLine(t2.sourceString); //v4.8
Console.WriteLine(t1 == t2); //true
```

**Comment by zms9110750** [comment]
Hello everyone. Although I don't know if it's useful for you. But perhaps you can take a look.
https://github.com/diegofrata/Generator.Equals
